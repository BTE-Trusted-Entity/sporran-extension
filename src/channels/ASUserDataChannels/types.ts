import { ICredential, KiltAddress } from '@kiltprotocol/types';
import { HexString } from '@polkadot/util/types';

import { DAppName } from '../AccessChannels/DAppName';
import { Origin } from '../AccessChannels/Origin';

export type ASUserDataInput = DAppName & {
  submitter: KiltAddress;
};

export type ASUserDataOriginInput = ASUserDataInput & Origin;

export interface ASUserDataOutput {
  createDidExtrinsic: HexString;
  credential: ICredential;
}
