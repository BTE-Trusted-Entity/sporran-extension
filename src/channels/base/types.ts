export interface ErrorFirstCallback<Output> {
  (error: null, output: Output): void;

  (error: Error): void;
}
