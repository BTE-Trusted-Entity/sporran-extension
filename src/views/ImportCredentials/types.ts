import { KiltAddress } from '@kiltprotocol/sdk-js';

export interface Import {
  fileName: string;
}

export interface SuccessfulImport extends Import {
  identityAddress: KiltAddress;
}

export interface FailedImport extends Import {
  error: 'invalid' | 'orphaned' | string;
}
