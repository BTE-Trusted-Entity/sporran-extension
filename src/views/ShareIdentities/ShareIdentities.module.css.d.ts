export const __esModule: true;
export const container: string;
export const heading: string;
export const details: string;
export const label: string;
export const origin: string;
export const choose: string;
export const identities: string;
export const identityLine: string;
export const avatar: string;
export const name: string;
export const disabled: string;
export const names: string;
export const namesInput: string;
export const buttonsLine: string;
export const reject: string;
export const submit: string;
export const errorTooltip: string;
