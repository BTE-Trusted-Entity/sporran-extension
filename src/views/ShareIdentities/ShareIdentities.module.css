.container {
  composes: container from '../../components/View/View.module.css';
  height: calc(100% - 1.5rem);
  text-align: left;
}

.heading {
  composes: screenHeadline from '../../components/Typography/Typography.module.css';
  margin-bottom: 2rem;
}

.details {
  width: 100%;
}

.label {
  composes: smallText from '../../components/Typography/Typography.module.css';
  margin: 0;
}

.origin {
  composes: inputText from '../../components/Typography/Typography.module.css';
  margin-block: 0.5rem;
}

.choose {
  composes: subline from '../../components/Typography/Typography.module.css';
}

.identities {
  composes: fullWidthContainer from '../../components/View/View.module.css';
  min-height: 3rem;
  flex-grow: 1;
  padding-block: 0.5rem;
  margin-block: 0.75rem;
}

.identityLine {
  composes: subline from '../../components/Typography/Typography.module.css';
  height: 2.5rem;
  display: flex;
  justify-content: start;
  align-items: center;
  margin-left: 1rem;
  cursor: pointer;
}

.avatar {
  composes: avatar from '../../components/Avatar/Avatar.module.css';
  height: 1.75rem;
  width: 1.75rem;
  background: none;
  margin-inline: 0.5rem;
}

.name {
  width: 24rem;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.disabled {
  composes: name;
  color: var(--color-grey-medium)
}

.names {
  composes: subline from '../../components/Typography/Typography.module.css';
  display: flex;
  gap: 0.5rem;
  cursor: pointer;
}

.namesInput {
  composes: checkbox from '../../components/Checkbox/Checkbox.module.css';
}

.buttonsLine {
  composes: buttonsLine from '../../components/Button/Button.module.css';
  margin-top: 0.5rem;
}

.reject {
  composes: buttonSecondary from '../../components/Button/Button.module.css';
}

.submit {
  composes: buttonPrimary from '../../components/Button/Button.module.css';
}

.errorTooltip {
  composes: tooltip from '../../components/ErrorTooltip/ErrorTooltip.module.css';
  top: 2.5rem;
  left: 75%;
}
