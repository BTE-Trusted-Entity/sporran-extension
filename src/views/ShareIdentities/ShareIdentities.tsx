import { FormEvent, useCallback, useState } from 'react';
import browser from 'webextension-polyfill';
import { pick } from 'lodash-es';

import * as styles from './ShareIdentities.module.css';

import { usePopupData } from '../../utilities/popups/usePopupData';
import {
  ShareIdentitiesOriginInput,
  ShareIdentitiesOutput,
} from '../../channels/ShareIdentitiesChannels/types';
import { backgroundShareIdentitiesChannel } from '../../channels/ShareIdentitiesChannels/backgroundShareIdentitiesChannel';
import { useIdentities } from '../../utilities/identities/identities';
import { Avatar } from '../../components/Avatar/Avatar';
import { isFullDid } from '../../utilities/did/did';

export function ShareIdentities() {
  const t = browser.i18n.getMessage;

  const identities = useIdentities().data;
  const { origin } = usePopupData<ShareIdentitiesOriginInput>();

  const [error, setError] = useState(false);
  const handleChange = useCallback(() => setError(false), []);

  const handleSubmit = useCallback(
    async (event: FormEvent<HTMLFormElement>) => {
      event.preventDefault();

      if (!identities) {
        return;
      }

      const formData = new FormData(event.currentTarget);
      const includeName = formData.has('names');
      const addresses = [...formData.keys()];
      const selected = Object.values(pick(identities, addresses));

      if (selected.length === 0) {
        setError(true);
        return;
      }

      const result = selected
        .filter(({ did }) => isFullDid(did))
        .map(({ name, did }) => ({
          did,
          ...(includeName && { name }),
        })) as ShareIdentitiesOutput;
      await backgroundShareIdentitiesChannel.return(result);
      window.close();
    },
    [identities],
  );

  const handleCancel = useCallback(async () => {
    await backgroundShareIdentitiesChannel.throw('Rejected');
    window.close();
  }, []);

  if (!identities) {
    return null; // storage data pending
  }

  return (
    <form className={styles.container} onSubmit={handleSubmit}>
      <h1 className={styles.heading}>{t('view_GetDidList_heading')}</h1>

      <section className={styles.details}>
        <p className={styles.label}>{t('view_GetDidList_origin')}</p>
        <p className={styles.origin}>{origin}</p>
      </section>

      <h2 className={styles.choose}>{t('view_GetDidList_choose')}</h2>

      <ul className={styles.identities}>
        {Object.values(identities).map((identity) => {
          const { name, did, address } = identity;
          const inactive = !isFullDid(did);
          return (
            <li key={address}>
              <label className={styles.identityLine}>
                <input
                  className={styles.namesInput}
                  type="checkbox"
                  name={address}
                  onChange={handleChange}
                  disabled={inactive}
                />
                <span />
                <Avatar identity={identity} className={styles.avatar} />
                <span className={inactive ? styles.disabled : styles.name}>
                  {inactive && t('view_GetDidList_noDID')}
                  {name}
                </span>
              </label>
            </li>
          );
        })}
      </ul>

      <p>
        <label className={styles.names}>
          <input className={styles.namesInput} type="checkbox" name="names" />
          <span />
          {t('view_GetDidList_names')}
        </label>
      </p>

      <p className={styles.buttonsLine}>
        <button onClick={handleCancel} type="button" className={styles.reject}>
          {t('common_action_cancel')}
        </button>
        <button type="submit" className={styles.submit}>
          {t('view_GetDidList_share')}
        </button>
        <output className={styles.errorTooltip} hidden={!error}>
          {t('view_GetDidList_choose')}
        </output>
      </p>
    </form>
  );
}
