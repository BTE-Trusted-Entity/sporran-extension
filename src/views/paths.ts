export { generatePath } from 'react-router-dom';

export const paths = {
  home: '/',
  settings: '/settings',
  access: '/access',
  identity: {
    base: '/identity/',
    add: '/identity/add',
    create: {
      start: '/identity/create',
      backup: '/identity/create/backup',
      verify: '/identity/create/verify',
      password: '/identity/create/password',
    },
    import: {
      start: '/identity/import',
      password: '/identity/import/password',
    },

    overview: '/identity/:address/:type(created|imported|pwreset)?',
    reset: {
      start: '/identity/:address/reset',
      password: '/identity/:address/reset/password',
    },
    send: {
      start: '/identity/:address/send',
      review: '/identity/:address/send/review',
      warning: '/identity/:address/send/warning',
    },
    receive: '/identity/:address/receive',
    credentials: {
      base: '/identity/:address/credentials',
      presentation: '/identity/:address/credentials/:hash/presentation',
    },
    remove: '/identity/:address/remove',
    did: {
      upgrade: {
        start: '/identity/:address/did/upgrade',
        kilt: '/identity/:address/did/upgrade/kilt',
        euro: '/identity/:address/did/upgrade/euro',
      },
      manage: {
        start: '/identity/:address/did',
        downgrade: {
          base: '/identity/:address/did/downgrade',
          warning: {
            web3name: '/identity/:address/did/downgrade/warning/web3name',
            credentials: '/identity/:address/did/downgrade/warning/credentials',
          },
          sign: '/identity/:address/did/downgrade/sign',
        },
        endpoints: {
          start: '/identity/:address/did/endpoints',
          edit: '/identity/:address/did/endpoints/:id?',
          sign: '/identity/:address/did/endpoints/sign',
        },
      },
      repair: '/identity/:address/did/repair',
    },
    web3name: {
      create: {
        base: '/identity/:address/web3name/create',
        info: '/identity/:address/web3name/create/info',
        form: '/identity/:address/web3name/create/form',
        choose: '/identity/:address/web3name/create/choose/:web3name',
        kilt: '/identity/:address/web3name/create/kilt/:web3name',
        euro: '/identity/:address/web3name/create/euro/:web3name',
      },
      manage: {
        start: '/identity/:address/web3name',
        remove: '/identity/:address/web3name/remove',
      },
    },
  },
  popup: {
    base: '/popup',
    import: '/popup/import',
    access: '/popup/access',
    claim: '/identity/:address/claim',
    save: '/popup/save',
    share: {
      start: '/popup/share',
      sign: '/popup/share/sign',
    },
    sign: '/popup/sign',
    signRaw: '/popup/signRaw',
    signDid: {
      start: '/identity/:address/did/sign',
      sign: '/identity/:address/did/sign/password',
      credentials: '/identity/:address/did/sign/credentials',
    },
    signDidExtrinsic: '/identity/:address/did/signExtrinsic',
    createDid: '/identity/:address/did/create',
  },
};
