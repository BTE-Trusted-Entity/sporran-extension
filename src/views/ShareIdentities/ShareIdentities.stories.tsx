import { Meta } from '@storybook/react';
import { JSX } from 'react';

import { PopupTestProvider } from '../../utilities/popups/PopupTestProvider';
import { ShareIdentitiesOriginInput } from '../../channels/ShareIdentitiesChannels/types';
import { paths } from '../paths';

import { ShareIdentities } from './ShareIdentities';

export default {
  title: 'Views/ShareIdentities',
  component: ShareIdentities,
} as Meta;

const input: ShareIdentitiesOriginInput = {
  dAppName: 'dApp',
  origin: 'https://example.org/foo',
};

export function Template(): JSX.Element {
  return (
    <PopupTestProvider path={paths.popup.shareIdentities} data={input}>
      <ShareIdentities />
    </PopupTestProvider>
  );
}
