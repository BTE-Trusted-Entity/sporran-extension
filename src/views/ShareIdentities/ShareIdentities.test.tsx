import { render } from '../../testing/testing';
import { PopupTestProvider } from '../../utilities/popups/PopupTestProvider';
import { paths } from '../paths';
import { ShareIdentitiesOriginInput } from '../../channels/ShareIdentitiesChannels/types';

import { ShareIdentities } from './ShareIdentities';

const mockInput: ShareIdentitiesOriginInput = {
  dAppName: 'dApp',
  origin: 'https://example.org/foo',
};

describe('ShareIdentities', () => {
  it('should render', async () => {
    const { container } = render(
      <PopupTestProvider path={paths.popup.shareIdentities} data={mockInput}>
        <ShareIdentities />
      </PopupTestProvider>,
    );

    expect(container).toMatchSnapshot();
  });
});
