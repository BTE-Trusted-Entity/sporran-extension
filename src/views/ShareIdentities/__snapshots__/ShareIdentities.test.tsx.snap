// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShareIdentities should render 1`] = `
<div>
  <form
    class="container"
  >
    <h1
      class="heading"
    >
      Share your Identities
    </h1>
    <section
      class="details"
    >
      <p
        class="label"
      >
        Request from:
      </p>
      <p
        class="origin"
      >
        https://example.org/foo
      </p>
    </section>
    <h2
      class="choose"
    >
      Choose at least one Identity with a DID to share
    </h2>
    <ul
      class="identities"
    >
      <li>
        <label
          class="identityLine"
        >
          <input
            class="namesInput"
            disabled=""
            name="4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1"
            type="checkbox"
          />
          <span />
          <div
            class="avatar"
          >
            Identicon
          </div>
          <span
            class="disabled"
          >
            [no DID] 
            Light DID Identity
          </span>
        </label>
      </li>
      <li>
        <label
          class="identityLine"
        >
          <input
            class="namesInput"
            disabled=""
            name="4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo"
            type="checkbox"
          />
          <span />
          <div
            class="avatar"
          >
            Identicon
          </div>
          <span
            class="disabled"
          >
            [no DID] 
            Identity with on-chain DID and lots of coins
          </span>
        </label>
      </li>
      <li>
        <label
          class="identityLine"
        >
          <input
            class="namesInput"
            disabled=""
            name="4pUVoTJ69JMuapNducHJPU68nGkQXB7R9xAWY9dmvUh42653"
            type="checkbox"
          />
          <span />
          <div
            class="avatar"
          >
            Identicon
          </div>
          <span
            class="disabled"
          >
            [no DID] 
            Identity which has a very very very very long name and might not even fit on the screen
          </span>
        </label>
      </li>
    </ul>
    <p>
      <label
        class="names"
      >
        <input
          class="namesInput"
          name="names"
          type="checkbox"
        />
        <span />
        Include names when sharing
      </label>
    </p>
    <p
      class="buttonsLine"
    >
      <button
        class="reject"
        type="button"
      >
        Cancel
      </button>
      <button
        class="submit"
        type="submit"
      >
        Share
      </button>
      <output
        class="errorTooltip"
        hidden=""
      >
        Choose at least one Identity with a DID to share
      </output>
    </p>
  </form>
</div>
`;
